fontforge/python.c

```diff
@@ -2441,169 +2476,98 @@ static PyObject *PyFFContour_InPlaceConcat( PyObject *_self, PyObject *_c2 ) {
+	PyFFContour_dealloc(c2);
     PyFFContour_ClearSpiros((PyFF_Contour *) self);
-Py_RETURN( self );
+    Py_RETURN( self );
 }
 
 static PyObject *PyFFContour_Index( PyObject *self, Py_ssize_t pos ) {
     PyFF_Contour *cont = (PyFF_Contour *) self;
     PyObject *ret;
 
-    if ( pos<0 || pos>=cont->pt_cnt ) {
+    if ( pos<-cont->pt_cnt || pos>=cont->pt_cnt ) {
 	PyErr_Format(PyExc_TypeError, "Index out of bounds");
-return( NULL );
+	return( NULL );
     }
+
+    if (pos < 0)
+	pos += cont->pt_cnt;
+
     ret = (PyObject *) cont->points[pos];
     Py_INCREF(ret);
 return( ret );
 }
 
 static int PyFFContour_IndexAssign( PyObject *self, Py_ssize_t pos, PyObject *val ) {
     PyFF_Contour *cont = (PyFF_Contour *) self;
-    PyObject *old;
+    PyFF_Point *old, *vpoint = NULL;
+    int i;
 
-    if ( !PyType_IsSubtype(&PyFF_PointType, Py_TYPE(val)) ) {
-	PyErr_Format(PyExc_TypeError, "Value must be a (FontForge) Point");
-return( -1 );
+    if ( val!=NULL && ( vpoint = PyFFPoint_Parse(val, false, false) ) == NULL ) {
+	PyErr_Format(PyExc_TypeError, "Unknown point format");
+	return( -1 );
     }
-    if ( pos<0 || pos>=cont->pt_cnt ) {
+    if ( pos<-cont->pt_cnt || pos>=cont->pt_cnt ) {
 	PyErr_Format(PyExc_TypeError, "Index out of bounds");
-return( -1 );
-    }
-    if ( cont->points[pos]->on_curve != ((PyFF_Point *) val)->on_curve &&
-	    !cont->is_quadratic ) {
-	PyErr_Format(PyExc_TypeError, "Replacement point must have the same on_curve setting as original in a cubic contour");
-return( -1 );
+	return( -1 );
     }
 
-    old = (PyObject *) cont->points[pos];
-    cont->points[pos] = (PyFF_Point *) val;
-    PyFFContour_ClearSpiros((PyFF_Contour *) self);
-    Py_DECREF( old );
-return( 0 );
-}
+    if (pos < 0)
+	pos += cont->pt_cnt;
 
-static PyObject *PyFFContour_Slice( PyObject *self, Py_ssize_t start, Py_ssize_t end ) {
-    PyFF_Contour *cont = (PyFF_Contour *) self;
-    PyFF_Contour *ret;
-    int len, i;
-
-    /* When passed in, start,end >= -cont->pt_cnt */
-    /* However, start,end can be arbitrarily large, particularly when the notation c[i:]
-	is used, in which case end is a very large value (max size of uint). */
-    if ( start<0 )
-	start += cont->pt_cnt;
-    if ( start>cont->pt_cnt )
-	start = cont->pt_cnt;
-    if ( end<0 )
-	end += cont->pt_cnt;
-    if ( end>cont->pt_cnt )
-	end = cont->pt_cnt;
-
-    if ( end<start )
-	len = end - start + cont->pt_cnt;
-    else
-	len = end - start;
+    old = cont->points[pos];
 
-    ret = (PyFF_Contour *)PyFF_ContourType.tp_alloc(&PyFF_ContourType, 0);
-    ret->is_quadratic = cont->is_quadratic;
-    ret->closed = false;
-    ret->pt_max = ret->pt_cnt = len;
-    ret->points = PyMem_New(PyFF_Point *,ret->pt_max);
-
-    if ( end<start ) {
-	for ( i=start; i<cont->pt_cnt; ++i )
-	    ret->points[i-start] = cont->points[i];
-	for ( i=0; i<end; ++i )
-	    ret->points[(cont->pt_cnt-start)+i] = cont->points[i];
+    if ( val==NULL ) {
+	for ( i=pos; i<=cont->pt_cnt-1; ++i )
```

Comment:
This check doesn't look right; should this condition be `i < cont->pt_cnt-1`? Otherwise wouldn't it overflow

Reply:
I think that's right. I'll check and test.

Reply:
This and the other one are now fixed in 60b59d575049e968294299a06838939b792918c5.